### PR TITLE
[mypy] Upgrade mypy and make suggested fixes

### PR DESCRIPTION
**Pull request by daniel-goldstein**
These changes are mostly because mypy now needs all `None` parameters to be explicit optionals, whereas before it would let something like `def foo(arg: str = None)` slide. The `type: ignore` on the azure client call is an azure library bug, I looked into the source for that function and there's a comment saying that the input is an `Optional[int]` but the type stubs that mypy is picking up say `int`.

the `check_untyped_defs` config flag tells mypy to typecheck the bodies of functions that do not have a type signature, which apparently it wasn't doing on its own.

### CODE REVIEW

**Comment by daniel-goldstein** [comment]
Sorry jumped the gun there, unassigning because there's still more linting to fix

File: hail/python/Makefile [hunk 1]
```diff
@@ -1,4 +1,4 @@
-PYTHON := python3
+PYTHON := python
```
Comment (jigold) [inline hail/python/Makefile:1]: fix.

**Review by jigold — APPROVED** [review]

**Review by jigold — CHANGES_REQUESTED** [review]
python -> python3

**Comment by jigold** [comment]
Looks like this needs a rebase now.

**Comment by jigold** [comment]
Unfortunately, needs a rebase again.

**Review by jigold — APPROVED** [review]

**Comment by jigold** [comment]
Approving in anticipation that the rebase is just fixing the pinned requirements. Can you re-request a review if you make other changes?

**Pull request merged**